### PR TITLE
fix(sidenav): update the way information stored in the tag

### LINKED ISSUE #995: Settings of left-side menu items for other users is changed when change of this menu for a current user
Order and available state of left-side menu items are changed for all users of domain after setting of left-side menu for any domain user.

Pre-conditions for config.json:
"allowReorderingSidebar": true;
In "sidebarOrder" there is several elements (Example: "sidebarOrder": ["setting", "ssh", "events", "volumes", "vms"]);

Steps:

1. Log in and go to left-side menu editing mode;
2. Make changes and exit left-side menu editing mode;
3. logout and log in as another user of this or other account.

Actual result: Settings of left-side menu items is changed for all users of domain;
Expected result: Settings of left-side menu items for each domain user is independent of other users settings;

Connected feature: sidebar_configure, sidebar_reorder

### PR DESCRIPTION
**Pull request by tamazlykar**
Fixes #995 #994 #992 
The problem was that only 255 characters can be stored in the tag. Previous storage structure was beyond this limitation and therefore there was an error.
Also, this PR includes some project improvements.

### CODE REVIEW

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 1]
```diff
@@ -174,4 +151,42 @@ export class AppSidebarComponent extends WithUnsubscribe()
...
+    if (
+      typeof order === 'undefined' ||
+      !Array.isArray(order) ||
+      order.length !== sidenavRoutes.length
+    ) {
+      return false;
+    }
+
+    return order.every(
+      _ =>
```
Comment (vanekbr) [inline src/app/core/components/sidenav/sidenav.component.ts]: Please, use human-readable semantic names here and in the next method

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 2]
```diff
@@ -170,8 +147,43 @@ export class AppSidebarComponent extends WithUnsubscribe()
...
       if (defaultOrder) {
-        this.routes = this.routes.filter(route => defaultOrder.some(_ => _.toUpperCase() === route.id));
+        this.routes = this.routes.filter(route =>
+          defaultOrder.some(orderElement => orderElement.toUpperCase() === route.id));
       }
     }
   }
+
+  private convertResponseToNavigationItemArray(response: string): NavigationItem[] {
+    let unconvertedArray = [];
```
Comment (zolotyx) [inline src/app/core/components/sidenav/sidenav.component.ts]: get rid of mutation

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 3]
```diff
@@ -170,8 +147,43 @@ export class AppSidebarComponent extends WithUnsubscribe()
...
+    return unconvertedArray.map(element => {
+      const id: string = Object.keys(element)[0];
+      const enabled: boolean = element[id];
+      return { id, enabled };
+    });
+  }
+
+  private validateNavigationOrder(order: NavigationItem[]): boolean {
+    if (
+      typeof order === 'undefined' ||
```
Comment (zolotyx) [inline src/app/core/components/sidenav/sidenav.component.ts]: is it possible to pass an undefined or non-array argument into this method?

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 4]
```diff
@@ -110,9 +93,8 @@ export class AppSidebarComponent extends WithUnsubscribe()
       this.hasChanges = false;
       this.updatingOrder = true;
 
-      const newOrder: Array<NavigationItem> = this.routes.map(({ id, enabled }) => ({
-        id,
-        enabled
+      const newOrder: Array<{ [key: string]: boolean }> = this.routes.map(({ id, enabled }) => ({
```
Comment (zolotyx) [inline src/app/core/components/sidenav/sidenav.component.ts]: a very controversial way of storing an order. please think about something different

**Comment by tamazlykar** [comment]
Please check changes provided
/cc @zolotyx @vanekbr

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 5]
```diff
@@ -170,8 +145,48 @@ export class AppSidebarComponent extends WithUnsubscribe()
...
+    return order.every(el =>
+        el.enabled != null && el.id != null && !!sidenavRoutes.find(route => route.id === el.id));
+  }
+
+  private navigationPredicate(order: NavigationItem[]) {
+    return (a: NavigationItem, b: NavigationItem) =>
+      order.findIndex(el => el.id === a.id) - order.findIndex(el => el.id === b.id);
+  }
+
+  private stringifyMenuState(routes: NavigationItem[]): string {
```
Comment (zolotyx) [inline src/app/core/components/sidenav/sidenav.component.ts]: use reduce + join

File: src/app/core/components/sidenav/sidenav.component.ts [hunk 6]
```diff
@@ -170,8 +145,48 @@ export class AppSidebarComponent extends WithUnsubscribe()
...
+    let menuState = '';
+    for (const route of routes) {
+      menuState += `${route.id}:${+route.enabled};`
+    }
+    return menuState;
+  }
+
+  private parseMenuState(menuStateString: string): NavigationItem[] {
```
Comment (zolotyx) [inline src/app/core/components/sidenav/sidenav.component.ts]: remove mutation. use split, filter, map

**Pull request merged**